### PR TITLE
Assign multiple distribution sets to a target via mgmt api

### PR DESCRIPTION
**Pull request by StefanKlt**
Enable the /rest/v1/targets/{targetId}/assignedDS POST endpoint to also accept lists of distribution sets.
When 'multi assignments' is enabled it is now possible to assign multiple distribution sets to a single target with one request.

### CODE REVIEW

**Comment by laverman** [comment]
@StefanKlt please verify the commit history and double check the signed-offs.

**Comment by schabdo** [comment]
@hawkbit-bot verify please

File: hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/QuotaManagement.java [hunk 1]
```diff
@@ -80,10 +80,10 @@
     int getMaxTargetsPerRolloutGroup();
 
     /**
-     * @return the maximum number of targets which for a manual distribution set
-     *         assignment
+     * @return the maximum number of actions resulting from a manual assignment
+     *         of distribution sets and targets
      */
-    int getMaxTargetsPerManualAssignment();
+    int getMaxResultingActionsPerManualAssignment();
```
Comment (StefanKlt) [inline hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/QuotaManagement.java]: also accept old property and map it
check quota docu

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaDeploymentManagement.java [hunk 2]
```diff
@@ -150,80 +157,96 @@ protected JpaDeploymentManagement(final EntityManager entityManager, final Actio
...
-    @Retryable(include = {
-            ConcurrencyFailureException.class }, maxAttempts = Constants.TX_RT_MAX, backoff = @Backoff(delay = Constants.TX_RT_DELAY))
-    public DistributionSetAssignmentResult assignDistributionSet(final long dsID,
-            final Collection<TargetWithActionType> targets, final String actionMessage) {
+    private List<DeploymentRequest> validateRequestForAssignments(
+            final Collection<DeploymentRequest> deploymentRequests) {
+        final List<DeploymentRequest> modifiableListOfRequests = new ArrayList<>(deploymentRequests);
+        if (!isMultiAssignmentsEnabled()) {
+            removeDuplicatesFromCollection(modifiableListOfRequests);
+            if (isCausingMultiassignment(modifiableListOfRequests)) {
```
Comment (StefanKlt) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaDeploymentManagement.java]: rename to check...

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaDeploymentManagement.java [hunk 3]
```diff
@@ -337,37 +363,31 @@ private JpaDistributionSet getAndValidateDsById(final Long dsID) {
...
     }
 
-    private void checkQuotaForAssignment(final Collection<TargetWithActionType> targetsWithActionType,
-            final JpaDistributionSet distributionSet) {
-        // enforce the 'max targets per manual assignment' quota
-        if (!targetsWithActionType.isEmpty()) {
-            assertMaxTargetsPerManualAssignmentQuota(distributionSet.getId(), targetsWithActionType.size());
+    private void checkQuotaForAssignment(final Collection<DeploymentRequest> deploymentRequests) {
+        if (!deploymentRequests.isEmpty()) {
+            enforceMaxActionsPerReqest(deploymentRequests.size());
```
Comment (StefanKlt) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaDeploymentManagement.java]: typo

File: hawkbit-rest/hawkbit-rest-docs/src/main/asciidoc/targets-api-guide.adoc [hunk 4]
```diff
@@ -601,6 +601,60 @@ include::../errors/429.adoc[]
 |===
 
 
+== POST /rest/v1/targets/{targetId}/assignedDS (assign multiple distribution sets)
+
+=== Implementation Notes
+
+Handles the POST request for assigning multiple distribution sets to a specific target (only allowed id 'multi assignments' is enabled). Required Permission: READ_REPOSITORY and UPDATE_TARGET
+
+=== Asssin distribution sets to target
```
Comment (StefanKlt) [inline hawkbit-rest/hawkbit-rest-docs/src/main/asciidoc/targets-api-guide.adoc]: typo

**Comment by schabdo** [comment]
@hawkbit-bot verify please

File: hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/DeploymentManagement.java [hunk 5]
```diff
@@ -48,71 +48,12 @@
...
-     * Assigns the given set of {@link DistributionSet} entities to the given
-     * {@link Target}s using the specified {@link ActionType} and
-     * {@code forcetime}.
-     *
-     * @param dsIDs
+     * @param deploymentRequests
      *            the set of IDs of the distribution sets to assign
-     * @param targets
-     *            a list of all targets and their action type
+     *
```
Comment (stefbehl) [inline hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/DeploymentManagement.java]: javadoc misleading... should be sth like "the set of deployment requests specifying the distribtution set assignments..."

File: hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/DeploymentManagement.java [hunk 6]
```diff
@@ -126,26 +67,29 @@ DistributionSetAssignmentResult assignDistributionSet(long dsID,
...
-     * their IDs with a specific {@link ActionType} and an action message.
+     * Assigns {@link DistributionSet}s to {@link Target}s according to the
+     * {@link DeploymentRequest}.
      *
-     * @param dsID
-     *            the ID of the distribution set to assign
-     * @param targets
-     *            a list of all targets and their action type
+     * @param deploymentRequests
+     *            the set of IDs of the distribution sets to assign
```
Comment (stefbehl) [inline hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/DeploymentManagement.java]: same as above... javadoc misleading...

File: hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/exception/MultiassignmentIsNotEnabledException.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+import org.eclipse.hawkbit.exception.AbstractServerRtException;
+import org.eclipse.hawkbit.exception.SpServerError;
+
+/**
+ * This exception is thrown if an operation requires multiassignments, but the
+ * feature is not enabled.
+ * 
+ */
+public class MultiassignmentIsNotEnabledException extends AbstractServerRtException {
```
Comment (stefbehl) [inline hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/exception/MultiassignmentIsNotEnabledException.java]: Minor: I'd prefer "MultiAssignment..." ...

File: hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/model/DeploymentRequest.java [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+package org.eclipse.hawkbit.repository.model;
+
+import org.eclipse.hawkbit.repository.exception.InvalidMaintenanceScheduleException;
+import org.eclipse.hawkbit.repository.model.Action.ActionType;
+
+/**
+ * A custom view on assigning a {@link DistributionSet} to a {@link Target}.
+ *
+ */
+public class DeploymentRequest {
```
Comment (stefbehl) [inline hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/model/DeploymentRequest.java]: Is there sth like a DeploymentRequestBuilder to create a DeploymentRequest based on a static builder pattern? The DeploymentManagement interface could offer a method "deploymentRequest() which returns a DeploymentRequestBuilder object...

Not sure how often DeploymentRequest objects need to be created... depending on that a builder pattern might be helpful...

File: hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/model/TargetWithActionType.java [hunk 9]
```diff
@@ -122,4 +141,66 @@ public String toString() {
...
+        result = prime * result + ((controllerId == null) ? 0 : controllerId.hashCode());
+        result = prime * result + (int) (forceTime ^ (forceTime >>> 32));
+        result = prime * result + ((maintenanceSchedule == null) ? 0 : maintenanceSchedule.hashCode());
+        result = prime * result + ((maintenanceWindowDuration == null) ? 0 : maintenanceWindowDuration.hashCode());
+        result = prime * result + ((maintenanceWindowTimeZone == null) ? 0 : maintenanceWindowTimeZone.hashCode());
+        return result;
+    }
+
+    @Override
+    public boolean equals(final Object obj) {
```
Comment (stefbehl) [inline hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/model/TargetWithActionType.java]: Check if Objects::equals could help here to simplify this implementation...
```
public static boolean equals(Object a, Object b) {
    return (a == b) || (a != null && a.equals(b));
}
```

File: MIGRATION.md [hunk 10]
```diff
@@ -17,4 +17,8 @@
 - Annotation `org.eclipse.hawkbit.ddi.resource.EnableDirectDeviceApi` have changed to `org.eclipse.hawkbit.ddi.annotation.EnableDdiApi`
 
 ### Renamed maven modules
-- Module hawkbit-mgmt-api-client have changed to hawkbit-example-mgmt-simulator
\ No newline at end of file
+- Module hawkbit-mgmt-api-client have changed to hawkbit-example-mgmt-simulator
```
Comment (stefbehl) [inline MIGRATION.md]: "have changed" should be "has changed"... also above...

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaDeploymentManagement.java [hunk 11]
```diff
@@ -150,80 +157,88 @@ protected JpaDeploymentManagement(final EntityManager entityManager, final Actio
...
-                onlineDsAssignmentStrategy);
-        onlineDsAssignmentStrategy.sendDeploymentEvents(result);
-        return result;
+    private static Map<Long, List<TargetWithActionType>> convertRequest(
+            final Collection<DeploymentRequest> deploymentRequests) {
+        return deploymentRequests.stream().collect(Collectors.groupingBy(DeploymentRequest::getDistributionSetId,
+                Collectors.mapping(DeploymentRequest::getTargetWithActionType, Collectors.toList())));
+    }
+
+    private static void checkIfCausesMultiassignment(final Collection<DeploymentRequest> deploymentRequests) {
```
Comment (stefbehl) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaDeploymentManagement.java]: Semantics of this method seems to be to check if the deployment would actually require Multi-Assignments, right? If yes, I'd rename the method to sth like "checkIfRequiresMultiAssignment"...

File: hawkbit-ui/src/main/resources/messages.properties [hunk 12]
```diff
@@ -516,6 +516,7 @@ message.dist.type.discard.success = All Distribution Types are discarded success
 message.dist.discard.success = All Distributions are discarded successfully !
 message.assign.discard.success = All assignments are discarded successfully !
 message.target.ds.assign.success = Assignment saved successfully !
+message.target.ds.multiassign.error = Can not assign multiple distribution sets to a target!
```
Comment (stefbehl) [inline hawkbit-ui/src/main/resources/messages.properties]: Should be "Cannot"

File: hawkbit-rest/hawkbit-mgmt-resource/src/test/java/org/eclipse/hawkbit/mgmt/rest/resource/MgmtTargetResourceTest.java [hunk 13]
```diff
@@ -1866,4 +1868,69 @@ public void searchDistributionSetMetadataRsql() throws Exception {
                 .andExpect(jsonPath("total", equalTo(1))).andExpect(jsonPath("content[0].key", equalTo("knownKey1")))
                 .andExpect(jsonPath("content[0].value", equalTo("knownValue1")));
     }
+
+    @Test
+    @Description("A request for assigning multiple DS to a rarget results in a Bad Request when multiassignment in disabled.")
```
Comment (stefbehl) [inline hawkbit-rest/hawkbit-mgmt-resource/src/test/java/org/eclipse/hawkbit/mgmt/rest/resource/MgmtTargetResourceTest.java]: "rarget" -> "target"

File: hawkbit-rest/hawkbit-mgmt-resource/src/test/java/org/eclipse/hawkbit/mgmt/rest/resource/MgmtDistributionSetResourceTest.java [hunk 14]
```diff
@@ -1273,4 +1272,63 @@ public void assignMultipleTargetsToDistributionSetAsDownloadOnly() throws Except
...
+        body.put(getAssignmentObject(targetId, MgmtActionType.SOFT));
+        body.put(getAssignmentObject(targetId, MgmtActionType.FORCED));
+
+        mvc.perform(post("/rest/v1/distributionsets/{ds}/assignedTargets", dsId).content(body.toString())
+                .contentType(MediaType.APPLICATION_JSON)).andDo(MockMvcResultPrinter.print())
+                .andExpect(status().isBadRequest());
+    }
+
+    @Test
+    @Description("Identical assignments in a single request are removed when multiassignment in disabled.")
```
Comment (stefbehl) [inline hawkbit-rest/hawkbit-mgmt-resource/src/test/java/org/eclipse/hawkbit/mgmt/rest/resource/MgmtDistributionSetResourceTest.java]: "in disabled" -> "is disabled"

**Pull request merged**